app/javascript/components/workflows/professor/exams/new/editor/components/questions/Matching.tsx

```diff
@@ -58,254 +63,261 @@ const ChooseRightAnswer: React.FC<{
-    newValues.splice(to, 0, fromOpt);
-    const newValue = { ...value };
-    Object.keys(newValue).forEach((key) => {
-      if (newValue[key] === from) {
-        newValue[key] = to;
-      } else if (newValue[key] === to) {
-        newValue[key] = to - 1;
-      }
-    });
-    return makeChangeAction({ ...info, values: newValues }, newValue);
-  };
+const renderPrompt = ({
+  numChoices,
+}) => (
+  member: string,
+  index: number,
+  fields: FieldArrayFieldsProps<HTMLVal>,
+) => (
+  <OnePrompt
+    // eslint-disable-next-line react/no-array-index-key
+    key={index}
+    valueNum={index}
+    numChoices={numChoices}
+    memberName={member}
+    enableDown={index + 1 < fields.length}
+    moveDown={(): void => {
+      fields.move(index, index + 1);
+    }}
+    moveUp={(): void => {
+      fields.move(index, index - 1);
+    }}
+    remove={(): void => {
+      fields.remove(index);
+    }}
+  />
+);
 
+const EditColName: React.FC<WrappedFieldProps & {
+  defaultLabel: string;
+}> = (props) => {
+  const {
+    input,
+    defaultLabel,
+  } = props;
+  const {
+    value,
+    onChange,
+  } = input;
+  return (
+    <CustomEditor
+      className="bg-white"
+      theme="bubble"
+      value={value.value ?? defaultLabel}
+      onChange={(newVal) => onChange({
+        type: 'HTML',
+        value: newVal,
+      })}
+    />
+  );
+};
+
+const RenderPrompts: React.FC<WrappedFieldProps> = (props) => {
+  const { input } = props;
+  const { length } = input.value;
+  const renderPrompts = React.useCallback(renderPrompt({
+    numChoices: length,
+  }), [length]);
+  return (
+    <FieldArray
+      name="prompts"
+      component={EditHTMLs}
+      renderOptions={renderPrompts}
+    />
+  );
+};
 
-  const setColA = (newColA: HTMLVal): void => {
-    onChange({
-      ...info,
-      promptsLabel: newColA,
-    }, value);
-  };
-  const setColB = (newColB: HTMLVal): void => {
-    onChange({
-      ...info,
-      valuesLabel: newColB,
-    }, value);
-  };
-  const updateAnswer = (index: number, event: React.ChangeEvent<{ value: number }>): void => {
-    const val = event.target.value;
-    const ret = { ...value };
-    ret[index] = val;
-    onChange(info, ret);
-  };
-  const updatePrompt = (index: number, newPrompt: HTMLVal): void => {
-    const newPrompts = [...prompts];
-    newPrompts[index] = newPrompt;
-    onChange({ ...info, prompts: newPrompts }, value);
-  };
-  const updateValue = (index: number, newValue: HTMLVal): void => {
-    const newValues = [...values];
-    newValues[index] = newValue;
-    onChange({ ...info, values: newValues }, value);
-  };
+const RenderValues: React.FC<WrappedFieldProps> = (props) => {
+  const { input } = props;
+  const {
+    value,
+    onChange,
+  }: {
+    value: MatchingPromptWithAnswer[],
+    onChange: (newVal: MatchingPromptWithAnswer[]) => void;
+  } = input;
+  const changeAllAnswers = useCallback((oldVal: number, newVal: number, swap = true): void => {
```

Comment:
This function is definitely confusing.  Some comments would be helpful to explain what it's doing...

Reply:
does that help?

Reply:
It explains what your code is doing, but now I'm not convinced it's right.  Basically we need to figure out whether we're _swapping_ two items, or _moving_ one item to a new index.  In the code I'd originally written, I was _moving_ an item from one index to another, so everything between those two indices (inclusive) would get shifted up or down by one.  It looks like your code is swapping two items, but then I'm not following why there's an optional `swap` parameters.

Reply:
If you move a matching answer from index 1 to index 2, all of the prompts that used to have 1 for an answer need to now have 2, and all of the prompts that used to have 2 for an answer need to now have 1 (they swap).

Swap is false for removing - if I remove at index 2, I want every prompt with answer 2 to be changed to -1, leave previous -1 answers alone, and any answer greater than 2 needs to be subtracted by 1 to account for the removed answer.

Reply:
Those sound like two completely different functions -- one for swapping and one for deleting.  Seems to me that `renderValue` should take in two callbacks, instead: `swapAnswersAt(from, to)` and `deleteAnswer(index)`

Reply:
Yes, that's better